cluster-autoscaler/simulator/scheduler_based_predicates_checker.go

```diff
@@ -0,0 +1,277 @@
+	)
+
+	if err != nil {
+		return nil, fmt.Errorf("couldn't create scheduler framework; %v", err)
+	}
+
+	checker := &SchedulerBasedPredicateChecker{
+		framework:                framework,
+		delegatingSharedLister:   sharedLister,
+		informerBasedShardLister: NewEmptySnapshot(),
+	}
+
+	// this MUST be called after all the informers/listers are acquired via the
+	// informerFactory....Lister()/informerFactory....Informer() methods
+	informerFactory.Start(stop)
+
+	return checker, nil
+}
+
+// FitsAnyNode checks if the given pod can be placed on any of the given nodes.
+func (p *SchedulerBasedPredicateChecker) FitsAnyNode(clusterSnapshot ClusterSnapshot, pod *apiv1.Pod, nodeInfos map[string]*scheduler_nodeinfo.NodeInfo) (string, error) {
+	if clusterSnapshot == nil {
+		return "", fmt.Errorf("ClusterSnapshot not provided")
+	}
+	if nodeInfos != nil {
+		klog.Errorf("clusterSnapshot and nodeInfos are mutually exclusive!!!!")
+	}
+	return p.fitsAnyNode(clusterSnapshot, pod)
+}
+
+func (p *SchedulerBasedPredicateChecker) fitsAnyNode(clusterSnapshot ClusterSnapshot, pod *apiv1.Pod) (string, error) {
+	var nodeInfosList []*scheduler_nodeinfo.NodeInfo
+
+	nodeInfosList, err := clusterSnapshot.NodeInfos().List()
+	if err != nil {
+		// TODO(scheduler_framework_integration) distinguish from internal error and predicate error
+		klog.Errorf("Error obtaining nodeInfos from schedulerLister")
+		return "", fmt.Errorf("error obtaining nodeInfos from schedulerLister")
+	}
+	p.delegatingSharedLister.UpdateDelegate(clusterSnapshot)
+	defer p.delegatingSharedLister.ResetDelegate()
+	state := scheduler_framework.NewCycleState()
+	preFilterStatus := p.framework.RunPreFilterPlugins(context.TODO(), state, pod)
+	if !preFilterStatus.IsSuccess() {
+		return "", fmt.Errorf("error running pre filter plugins for pod %s; %s", pod.Name, preFilterStatus.Message())
+	}
+
+	for _, nodeInfo := range nodeInfosList {
+		// Be sure that the node is schedulable.
+		if nodeInfo.Node().Spec.Unschedulable {
+			continue
+		}
+
+		filterStatuses := p.framework.RunFilterPlugins(context.TODO(), state, pod, nodeInfo)
+		ok := true
+		for _, filterStatus := range filterStatuses {
+			if !filterStatus.IsSuccess() {
+				ok = false
+				break
+			}
+		}
+		if ok {
+			return nodeInfo.Node().Name, nil
+		}
+	}
+	return "", fmt.Errorf("cannot put pod %s on any node", pod.Name)
+}
+
+// CheckPredicates checks if the given pod can be placed on the given node.
+func (p *SchedulerBasedPredicateChecker) CheckPredicates(clusterSnapshot ClusterSnapshot, pod *apiv1.Pod, nodeInfo *scheduler_nodeinfo.NodeInfo) *PredicateError {
+	if clusterSnapshot == nil {
+		return NewPredicateError(InternalPredicateError, "", "ClusterSnapshot not provided", nil, emptyString)
+	}
+	return p.checkPredicates(clusterSnapshot, pod, nodeInfo.Node().Name)
+}
+
+func (p *SchedulerBasedPredicateChecker) checkPredicates(clusterSnapshot ClusterSnapshot, pod *apiv1.Pod, nodeName string) *PredicateError {
```

Comment:
Is there any reason to have a separate CheckPredicates and checkPredicates functions after you removed the legacy implementation?

Reply:
Not really - will merge in